src/protocols/bdx/AsyncTransferFacilitator.cpp

```diff
@@ -0,0 +1,154 @@
+    // transfer could necessitate a response if they are received at the wrong time.
+    // For this reason, it is left up to the application logic to call ExchangeContext::Close() when it has determined that the
+    // transfer is finished.
+    ec->WillSendMessage();
+
+    // Get the next output event and send it to the delegate.
+    TransferSession::OutputEvent outEvent;
+    do
+    {
+        mTransfer.GetNextAction(outEvent);
+
+        if (outEvent.EventType == TransferSession::OutputEventType::kTransferTimeout)
+        {
+            OnResponseTimeout(ec);
+        }
+        else
+        {
+            HandleAsyncTransferSessionOutput(outEvent);
+        }
+    } while (outEvent.EventType != TransferSession::OutputEventType::kNone);
+
+    return err;
+}
+
+Messaging::ExchangeContext * AsyncTransferFacilitator::GetExchangeContext()
+{
+    if (mExchange)
+    {
+        return mExchange.Get();
+    }
+    return nullptr;
+}
+
+void AsyncTransferFacilitator::OnExchangeClosing(chip::Messaging::ExchangeContext * ec)
+{
+    ChipLogDetail(BDX, "%s, ec: " ChipLogFormatExchange, __FUNCTION__, ChipLogValueExchange(ec));
+    if (mExchange)
+    {
+        mExchange.Release();
+    }
+}
+
+void AsyncTransferFacilitator::OnResponseTimeout(Messaging::ExchangeContext * ec)
+{
+    ChipLogDetail(BDX, "%s, ec: " ChipLogFormatExchange, __FUNCTION__, ChipLogValueExchange(ec));
+    if (mExchange)
+    {
+        mExchange.Release();
+    }
+}
+
+CHIP_ERROR AsyncResponder::PrepareForTransfer(Messaging::ExchangeContext * exchangeCtx, TransferRole role,
+                                              BitFlags<TransferControlFlags> xferControlOpts, uint16_t maxBlockSize,
+                                              System::Clock::Timeout timeout)
+{
+    VerifyOrReturnError(exchangeCtx != nullptr, CHIP_ERROR_INCORRECT_STATE);
+
+    ReturnErrorOnFailure(mTransfer.WaitForTransfer(role, xferControlOpts, maxBlockSize, timeout));
+
+    VerifyOrReturnError(!mExchange, CHIP_ERROR_INCORRECT_STATE);
+    mExchange.Grab(exchangeCtx);
+
+    return CHIP_NO_ERROR;
+}
+
+void AsyncResponder::ResetTransfer()
+{
+    if (mExchange)
+    {
+        mExchange.Release();
+    }
+    mTransfer.Reset();
+}
+
+bdx::StatusCode AsyncResponder::GetBdxStatusCodeFromChipError(CHIP_ERROR err)
+{
+    if (err == CHIP_ERROR_INCORRECT_STATE)
+    {
+        return bdx::StatusCode::kUnexpectedMessage;
+    }
+    if (err == CHIP_ERROR_INVALID_ARGUMENT)
+    {
+        return bdx::StatusCode::kBadMessageContents;
+    }
+    return bdx::StatusCode::kUnknown;
+}
+
+void AsyncResponder::NotifyEventHandledWithError(CHIP_ERROR error)
+{
+    ChipLogDetail(BDX, "%s : error %d", __FUNCTION__, error.AsInteger());
+    if (error != CHIP_NO_ERROR)
+    {
+        mTransfer.AbortTransfer(GetBdxStatusCodeFromChipError(error));
+    }
+
+    // Get the next output event  and send it to the delegate.
```

Comment:
This is a copy/paste of the code we had in OnMessageReceived, right?

It seems to me like:

1.  That code should live in one helper function.
2. Depending on our answer to whether it's OK to call this method synchronously while processing an event, we may want a boolean that we set while processing events that short-circuits calls to the "process an event" function so we don't do silly recurse-while-iterating things.

Reply:
yes it is. N yes i think it can be called synchronously while processing an event. I did plan to clean this up but didn't get to it for what its worth :).

Reply:
ok. so this might not be applicable anymore. the code in NotifyEventHandled should only handle MsgToSend for cases where the handling is done asynchronously in the subclass for e.g calling AcceptTransfer, PrepareBlock in the completionHandler and the AsyncFacilitator needs to know so it can get the next action which should be a MsgToSend. I ll let you take a look at the new code and decide if this is needed.